### DISCUSSION TITLE
Parse pre-tokenized text

### DISCUSSION THREAD

**Discussion by haocheng6** [explosion/spaCy#7813] [Help: Coding & Implementations] [answered]
I already have some pre-tokenized text, but I want to do everything that `nlp("some string")` can do except tokenization. I found this in the [documentation](https://spacy.io/usage/linguistic-features#own-annotations) that I can first manually create a Doc object and then do the rest job:
> Once you have a Doc object, you can write to its attributes to set the part-of-speech tags, syntactic dependencies, named entities and other attributes.

However, I still cannot find the right way to do the rest job in the documentation. Could anyone give me some tips? Thanks!

**Comment by polm** [comment, marked as answer]
The documentation you found is about setting attributes you already have, not using the existing model. That's useful if you're creating gold documents based on an existing corpus to create training data, for example. 

To use the pretrained models with pre-tokenized text, you can just replace the tokenizer with a function that takes a string and returns a Doc. [From the docs](https://spacy.io/usage/linguistic-features#custom-tokenizer:~:text=To%20overwrite%20the%20existing%20tokenizer%2C%20you,a%20text%20and%20returns%20a%20Doc.): 

```
nlp = spacy.blank("en")
nlp.tokenizer = my_tokenizer
```

So you can just load a model as usual, replace the tokenizer with a tokenizer that reads your string input and returns a Doc. 

Inside your tokenizer you would do something like this:

```
words = ... # get your list of pre-tokenized words
doc = Doc(nlp.vocab, words=words)
```

You may want to set other attributes you have access to, like spaces around words or sentence starts. Look at the [docs for the Doc class][doc] to see what your options are there.

[doc]: https://spacy.io/api/doc#init

That said, I would be cautious about doing this without training new models. The models depend on the tokenizer, so seemingly subtle changes could have weird effects. For example, if your tokenizer treats "hot-dog" as one token, while spaCy treats it as three, "hot-dog" will be an unk to the spaCy models, and won't have vectors and may get weird results.

**Reply by adrianeboyd** [reply]
See also how to modify the existing tokenizer to get a tokenizer that splits on single spaces: https://github.com/explosion/spaCy/discussions/7737#discussioncomment-599308

Like Paul already pointed out, the model performance may be degraded.
